lms/djangoapps/discussion/rest_api/api.py

```diff
@@ -1250,3 +1254,82 @@ def delete_comment(request, comment_id):
         comment_deleted.send(sender=None, user=request.user, post=cc_comment)
     else:
         raise PermissionDenied
+
+
+def _get_course_discussion_user_stats(
+    course_key: CourseKey,
+    order_by: UserOrdering = UserOrdering.BY_FLAGS,
+) -> List[Dict[str, Union[str, int]]]:
+    """
+    Get course user discussion stats.
+
+    Args:
+        course_key (CourseKey): course key for course to fetch discussion stats for.
+        order_by (UserOrdering): the ordering to use for sorting data
+
+    Returns:
+        Sorted user stat data for the specified course
+    """
+    course_stats = get_course_user_stats(course_key)
+
+    def key_func(user):
+        key = (user["threads"], user["responses"], user["replies"])
+        if order_by == UserOrdering.BY_FLAGS:
+            key = (user["active_flags"], user["inactive_flags"]) + key
+        return key
+
+    return sorted(course_stats.values(), key=key_func, reverse=True)
+
+
+def get_course_discussion_user_stats(
+    request,
+    course_key_str: str,
+    page: int,
+    page_size: int,
+    order_by: UserOrdering = None,
+) -> Dict:
+    """
+    Get paginated course discussion stats for users in the course.
+
+    Args:
+        request (Request): DRF request
+        course_key_str (str): course key string
+        page (int): Page number to fetch
+        page_size (int): Number of items in each page
+        order_by (UserOrdering): The ordering to use for the user stats
+
+    Returns:
+        Paginated data of a user's discussion stats sorted based on the specified ordering.
+
+    """
+    course_key = CourseKey.from_string(course_key_str)
+    if has_discussion_privileges(user=request.user, course_id=course_key):
+        order_by = order_by or UserOrdering.BY_FLAGS
+    else:
+        order_by = order_by or UserOrdering.BY_ACTIVITY
+        if order_by != UserOrdering.BY_ACTIVITY:
+            raise ValidationError({"order_by": ["Invalid value"]})
+    data = _get_course_discussion_user_stats(course_key, order_by)
+    num_items = len(data)
+    num_pages = ceil(num_items / page_size)
+    start_idx = (page - 1) * page_size
+    data = data[start_idx:start_idx + page_size]
+    user_ids = [user["author_id"] for user in data]
+    if not data and page != 1:
+        raise PageNotFoundError("Page not found (No results on this page).")
```

Comment:
Thanks, See here we are fetching (from mongo) all records & sorting them. Then we are choking them based on page number. It seems we would need to fetch/sort every time we need a next page which seems inefficient. 

Can we implement a request cache here or anything similar which can help us avoid fetch/sort (all records) for fetching the next pages? 

thanks.

Reply:
This was discussed on call, but adding here for the record. 

I've stayed away from implementing any caching, so we can get a good idea of the performance, and we know exactly where / how to cache. Caching will definitely be involved, but what form will it take?

1. Caching the request from cs_comment_service so when a user  clicks next we can just use the last results. This is simple, but will it be too much data for memcache etc? All records for all users for potentially hundreds of courses?
2. Caching the results post-sorting and pagination. This saves a lot of space since only one page is cached (or 2 since there are two sorts), and since sorting doesn't need to be done each time it saves on that as well. People will usually stick to early pages, but unless pagination is done quickly, the data may change between pages, so the requesting page n should refresh the cache of all pages up to n.
3. Running the mongo query async and saving it to the DB so the results are fetched directly from there. This will use pagincation from

Reply:
Thanks for thee detailed explaination. 

I am ok with this plan, can you please make sure the the pr https://github.com/edx/cs_comments_service/pull/352 is ready (is not in WIP/Draft state) so that we can merge that first?
thanks.